File: src/main/java/com/battletrident/games/ring/PhaseManager.java
package com.battletrident.games.ring;

public class PhaseManager {
	public final Ring[] RINGS = {
		new Ring(0, 400, 2, 0, 10),
		new Ring(1, 300, 60, 0.0001, 30),
		new Ring(2, 200, 60, 0.0001, 30),
		new Ring(3, 100, 60, 0.1, 30),
		new Ring(4, 1, 60, 1, -1),
	};
	private final RingUpdateEvent event = new RingUpdateEvent(RINGS[0], null);

	private int index = 0;

	public void notifyEnd() {
		event.setState(RingState.ENDED);
		event.callEvent();
	}
	public boolean hasMore() {
		return index + 1 < RINGS.length;
	}
	public Ring current() {
		return event.getRing();
	}
	public Ring next() {
		event.setRing(RINGS[++index]);
		event.callEvent();

		return event.getRing();
	}
	public Ring previous() {
		if (index > 0) {
			event.setRing(RINGS[--index]);
			event.callEvent();

			return event.getRing();
		} else throw new IndexOutOfBoundsException();
	}
	public void set(int i) {
		if (i >= 0 && i < RINGS.length) {
			event.setRing(RINGS[index = i]);
			event.callEvent();
		} else throw new IndexOutOfBoundsException();
	}
	public void clearTasks() {
		for (Ring ring : RINGS) {
			if (ring.ender != null)
				ring.ender.cancel();
			if (ring.starter != null)
				ring.starter.cancel();
		}
	}
}

File: src/main/kotlin/com/BattleTrident/games/ring/PhaseManager.kt
package com.battletrident.games.ring

import com.BattleTrident.games.ring.Ring
import com.BattleTrident.games.ring.RingState

class PhaseManager {
	private val RINGS = arrayOf(
		Ring(0, 300.0, 2, 0.0, 10),
		Ring(1, 200.0, 30, 0.0001, 30),
		Ring(2, 100.0, 20, 0.1, 30),
		Ring(3, 1.0, 60, 0.5, -1),
	)
	private val event = RingUpdateEvent(RINGS[0], null)

	private var index = 0

	fun notifyEnd() {
		event.setState(RingState.ENDED)
		event.callEvent()
	}

	fun hasMore(): Boolean {
		return index + 1 < RINGS.size
	}

	fun current(): Ring {
		return event.getRing()
	}

	fun next(): Ring {
		event.setRing(RINGS[++index])
		event.callEvent()

		return event.getRing()
	}

	fun set(i: Int) {
		if (i >= 0 && i < RINGS.size) {
			index = i
			event.setRing(RINGS[i])
			event.callEvent()
		} else throw IndexOutOfBoundsException()
	}

	fun clearTasks() {
		for (ring in RINGS) {
			ring.ender?.cancel()
			ring.starter?.cancel()
		}
	}
}